explainers/beforematch-spec-draft.md

```diff
@@ -0,0 +1,51 @@
+# beforematch event
+
+_These changes will likely go into the
+[find-in-page](https://html.spec.whatwg.org/multipage/interaction.html#find-in-page)
+section of the HTML spec_
+
+When a new active match is set, either by advancing through the match list or
+due to a new find-in-page request, a `beforematch` event is fired on a node
+identified by the active match [range
+start](https://dom.spec.whatwg.org/#concept-range-start). This process follows
+the following algorithm:
+
+1. Identify the match _M_ which will become the new active match. _M_'s DOM
+range must not be [collapsed](https://dom.spec.whatwg.org/#range-collapsed).
+
+2. At the next rendering opportunity, specifically step 12 of [Update the
+rendering](https://html.spec.whatwg.org/#rendering-opportunity) step:
+
+      2.1. If the range that constitutes match _M_ is
+        [collapsed](https://dom.spec.whatwg.org/#range-collapsed), restart the
+        algorithm at Step 1.
```

Comment:
We also restart the search next time starting after the range which got collapsed. The collapsed range is passed in to the next search the same way that selection is passed into find-in-page - when you select some text on a page and then start a find-in-page, it starts searching from (the beginning?) of the text the user selected.

If this is hard to define then maybe we could leave it out for now.

Reply:
Ok, I think I see what you're getting at. I'll add a note here.

Reply:
One of my patched in review right now also makes it so that during this circular finding process stops if it goes past the first range we matched by storing the first range and passing it through the AsyncScrollContext and the Find function itself. Also not sure if this is worth mentioning now or if it would be hard to define, but fyi.